core/domain/exp_domain.py

```diff
@@ -2176,6 +2176,52 @@ def _convert_states_v26_dict_to_v27_dict(cls, states_dict):
 
         return states_dict
 
+    @classmethod
+    def _convert_states_v27_dict_to_v28_dict(cls, states_dict):
+        """Converts from version 27 to 28. Version 28 adds content_id to
+        customization_arg for interactions which can have translations.
+
+        Args:
+            states_dict: dict. A dict where each key-value pair represents,
+                respectively, a state name and a dict used to initialize a
+                State domain object.
+
+        Returns:
+            dict. The converted states_dict.
+        """
+        translatble_interactions = [
+            'DragAndDropSortInput', 'ItemSelectionInput', 'MultipleChoiceInput']
+        for state_dict in states_dict.values():
+            if state_dict['interaction']['id'] in translatble_interactions:
+                customization_args = state_dict[
+                    'interaction']['customization_args']
+                new_values = []
+                content_ids = []
+                for index, value in enumerate(
+                        customization_args['choices']['value']):
+                    content_id = 'interaction_' + str(index + 1).zfill(5)
```

Comment:
There's a possible failure mode here. Suppose we already have v1, v2, v3 of an exploration in production with one content that needs a content ID. That piece of content has been changed several times so that v3's content is completely different from v1's (in fact a choice may have been deleted and then added later). We do a migration to v4 and introduce this content ID field, so the existing content is labelled 0.

Then we revert to an earlier version (v2, say). But now the content from that version will get labelled as 0 again and be associated with translations that are incorrect.

Is there any way to avoid this? We should have backend tests for this flow, too.

Reply:
@seanlip initially I was using this `content_id = 'interaction_' + utils.generate_random_string(5)` (similar to how I'm generating id in the frontend) but I realized that we have lots of backend test which follows this pattern:
- Old version exploration yaml.
- Do migration.
- Create a New version yaml.
- Expect that the newly migrated exploration should be equal to "New version yaml".

So for this kinda test, our migration should be deterministic I cannot use generate_random_string.

This is the only reason I have a deterministic pattern for generating the content id for all the places in the state, I'm not sure how I can avoid this. I'm not sure how important these kinda yaml tests are but I'm pretty sure in future we can have some random id in the state dict which will block this test. Do you have any suggestion (though I'll be going through this and see if I can find something)?

Reply:
I don't mind the migration being deterministic -- introducing randomness still adds a failure mode, i.e. chance of collision. But you might need to pass more information into the migration function in order to correctly handle the revert case I was talking about.